sdk/lib/opentelemetry/sdk/trace/tracer_provider.rb

```diff
@@ -38,7 +39,8 @@ def initialize
         def tracer(name = nil, version = nil)
           name ||= ''
           version ||= ''
-          @mutex.synchronize { @registry[Key.new(name, version)] ||= Tracer.new(name, version) }
+          resource = @resource.merge(OpenTelemetry::SDK::Resources::Resource.create('name' => name, 'version' => version))
```

Comment:
The name and version here should technically be the `InstrumentationLibrary`. We should be adding this to the span separately: https://github.com/open-telemetry/oteps/pull/84. Recording this as a Resource was an existing mistake we shouldn't carry forward.

Reply:
Incidentally, this work is going on in #264.

Reply:
I think I've found myself to be a little confused here and regarding your question here https://github.com/open-telemetry/opentelemetry-ruby/pull/264#pullrequestreview-413928944 

I believe my change to reflect what is in the specification:

>New Tracer instances are always created through a TracerProvider (see API). The name and version arguments supplied to the TracerProvider must be used to create a Resource instance which is stored on the created Tracer.

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#tracer-creation

So to try and answer your question on #264 I believe we need to keep the `library_resource` on the tracer so it is able to provide that information to the spans `getLibraryResource` method.

Reply:
I think that portion of the specification might be outdated.


The [SDK Resource Specification](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/resource/sdk.md) states:

> When used with distributed tracing, a resource can be associated with the TracerProvider when it is created. That association cannot be changed later. When associated with a TracerProvider, all Spans produced by any Tracer from the provider MUST be associated with this Resource.

That seems to indicate that there is a single resource that is referenced through the TracerProvider, that will be attached to all spans it generates. At a bare minimum, this Resource will be the [telemetry library resource](https://github.com/open-telemetry/opentelemetry-specification/tree/master/specification/resource/semantic_conventions#telemetry-sdk). If there are additional resources detected for a cloud, container, or based on environment variables they will all be merged into a single resource. The way a resource will make it to the export pipeline is basically a reference that gets attached to: TracerProvider -> Tracer -> Span -> SpanData.

The name and version of an individual tracer belongs to the instrumentation library which will flow from Tracer -> Span -> SpanData. This is specified [here](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#obtaining-a-tracer).

My interpretation is that there is a single Resource per TracerProvider (with labels composed from many underlying resources) that is made available on spans and also an InstrumentationLibrary reference on each span.

The protos also seem to support this interpretation. 
See: 
* https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/trace/v1/trace.proto#L28-L45
* https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/common/v1/common.proto#L57-L62
* https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/resource/v1/resource.proto#L26-L34

Note is a `ResourceSpans` message that has repeated `InstrumentationLibrarySpans` that has repeated `Spans`.

I don't see where the `library_resource` fits in.

As you've pointed out, there is a discrepancy between the wording in the [SDK tracer creation spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#tracer-creation) and this interpretation. To my knowledge, this was the way things were before the instrumentation library was introduced and that part of the spec might be outdated. I'm piecing this together from conversations that have happened at the spec sig, other implementations I've seen, and the protos. I realize this is hardly a definitive answer. I'll try to write up a spec issue to get some clarification.

Reply:
Also relevant: https://github.com/open-telemetry/oteps/blob/master/text/0083-component.md

Reply:
Also: https://github.com/open-telemetry/opentelemetry-specification/issues/614

Reply:
Each `Tracer` should have an `InstrumentationLibrary` with the `name` and `version`. The `TracerProvider` should have a `Resource`. The `Resource` is attached to every [ETSR](https://github.com/open-telemetry/opentelemetry-proto/blob/28e2774262a651c4c2e701c3a05820c3e43ead71/opentelemetry/proto/collector/trace/v1/trace_service.proto#L38-L45) in OTLP (technically to every `ResourceSpans`, but in practice the `repeated` in ETSR will be always 1 for ETSRs produced by clients - larger batches may be produced by the collector). Things get more complicated for other protocols, and we’ll need to figure out the mapping. We’re not alone here - the Jaeger exporter from OTel Go isn’t quite in the right shape yet.

Reply:
@fbogsany that matches my understanding and after raising in issue on the sec SIG it seems to be the current state of things. Through this process, I discovered: https://github.com/open-telemetry/opentelemetry-proto/pull/150, which, if accepted, might bring this full circle. We would end up having per tracer Resources that contain the instrumentation library and process level information.

Reply:
Piecing it together, I believe I've adjusted things to follow what was discussed above.

- the configurator is initialized a resource containing the [telemetry sdk](https://github.com/robertlaurin/opentelemetry-ruby/blob/ec04cac756da0620b293594c162787ed75604e3f/sdk/lib/opentelemetry/sdk/configurator.rb#L30) labels.
- the configurator optionally accepts a resource which is [merged with the default resource](https://github.com/robertlaurin/opentelemetry-ruby/blob/ec04cac756da0620b293594c162787ed75604e3f/sdk/lib/opentelemetry/sdk/configurator.rb#L37-L39)
- the tracer provider is initialized with the [resource from the configurator](https://github.com/robertlaurin/opentelemetry-ruby/blob/ec04cac756da0620b293594c162787ed75604e3f/sdk/lib/opentelemetry/sdk/configurator.rb#L96-L98)
- the tracer provider can initialize a tracer [(accepting name and version as optional params)](https://github.com/robertlaurin/opentelemetry-ruby/blob/ec04cac756da0620b293594c162787ed75604e3f/sdk/lib/opentelemetry/sdk/trace/tracer_provider.rb#L38-L42)
- the tracer has an instrumentation_library containing [name and version](https://github.com/robertlaurin/opentelemetry-ruby/blob/ec04cac756da0620b293594c162787ed75604e3f/sdk/lib/opentelemetry/sdk/trace/tracer.rb#L25)
- when a tracer creates a new span it passes along the resource from the [tracer provider](https://github.com/robertlaurin/opentelemetry-ruby/blob/ec04cac756da0620b293594c162787ed75604e3f/sdk/lib/opentelemetry/sdk/trace/tracer.rb#L57-L58)

Reply:
>I don't see where the library_resource fits in.
https://github.com/robertlaurin/opentelemetry-ruby/blob/09aa570537bada624810e8725050ca221e39a19b/sdk/lib/opentelemetry/sdk/trace/span.rb#L239

Should this be renamed or removed entirely, we still need to associate spans with the resource correct?

Reply:
I think the library resource should be removed. I'm going to PR the change I think is needed separately. I'll re-read the PR, but I don't think you need to change anything in this PR.